### ISSUE TITLE
Does adjusting the number of classes of a pretrained model work?

### ISSUE THREAD

**Issue by simonruber** [ultralytics/yolov5#9079]
### Search before asking

- [x]  I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions. *

### Question

Hi everyone,

I'm a bit confused about how to properly load a pretrained model with an adjusted number of classes for training with a custom dataset.

On the [Load YOLOv5 from PyTorch Hub ⭐](https://github.com/ultralytics/yolov5/issues/36) page you've explained that one can adjust  the  number of classes in the pretrained model by using the following command. `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', classes=10)`

<img width="999" alt="Bildschirmfoto 2022-08-22 um 08 13 15" src="https://user-images.githubusercontent.com/5917496/185851461-b177aa78-2b56-46a1-9c43-081d2a746938.png">

When I do so, I can see that a model.yaml file is overwritten, but I do not know where this file is stored. 

Now, what actually confuses me about the number of classes, is that when I try to use this pretrained model in detection, without any further training. I see an error, that the model was trained with nc=80 and my data is incompatible with nc=13:

`AssertionError: ['yolov5s6.pt'] (80 classes) trained on different --data than what you passed (13 classes). Pass correct combination of --weights and --data that are trained together.`

I know that I can not expect any proper predictions since the last layers are initialized with random weights, but I was expecting that the model is compatible with the 13 classes dataset.

Is this behavior to be expected or am I doing something wrong here? 
Do I need to find and use the model.yaml file and is the only thing changed in there 'nc=13'?

**Comment by MartinPedersenpp** [comment]
From: https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading
> Training
To load a YOLOv5 model for training rather than inference, set autoshape=False. To load a model with randomly initialized weights (to train from scratch) use pretrained=False. You must provide your own training script in this case. Alternatively see our YOLOv5 [Train Custom Data Tutorial](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data) for model training.
>  model = torch.hub.load('ultralytics/yolov5', 'yolov5s', autoshape=False)  # load pretrained
>  model = torch.hub.load('ultralytics/yolov5', 'yolov5s', autoshape=False, pretrained=False)  # load scratch

For training I would suggest using train.py instead of using Torch hub, unless you want to write your own training script

**Comment by simonruber** [comment]
@MartinPedersenpp  thank you for your answer. 
The way I planned to train was to first download a pretrained model with the correct number of classes (via torch hub) and afterwards pass this pretrained model as `--weights yolov5s6_with_13_classes.pt` to train.py. 
Since I see no possibility to set the 'classes' argument in train.py directly, I did not find another way to make sure that the correct model is used. 

Or is this simply not necessary since in train.py the information of the data.yaml is available, which specifies the number of classes?

**Comment by MartinPedersenpp** [comment]
@simonruber 
Yes, with train.py you just supply it with the correct pretrained yolov5.pt depending on your use case.
Normally "s" or "m" is a good first bet.
Then you customize the data.yaml to fit your custom classes and when training, the 80 classes from the COCO dataset will be wiped, but the features will be used for transfer learning to easier catch on to the features of your dataset.
I am not entirely sure what you are trying to do but, for the easiest first experiment, I would recommend using the "Train on custom dataset" tutorial.

**Comment by simonruber** [comment]
@MartinPedersenpp  thanks again, for your answer. I read through the "Train on custom dataset" tut already. I was simply not sure about structure of the model output layers, and whether (if one only has 13 output classes) the "unused" 67 output nodes remain in the model as long as one does not specify the number of classes when downloading the pretrained model.

But it looks like I approached it a bit too complicated, and the train.py / data.yaml are already taking care of the correct number of classes and are adjusting the model architecture.
I eventually just found another answer about my issue which points in the same direction: https://github.com/ultralytics/yolov5/discussions/5837#discussioncomment-1727907
